File: dotnet/Rezel/Grammars/BashGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class BashGrammar
{
    private static readonly NodePropSource BashHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["while do done until for in case esac if then elif else fi"] = Tags.ControlKeyword,
        ["IORedirect"] = Tags.Operator,
        ["&& || |"] = Tags.LogicOperator,
        ["= +="] = Tags.Operator,
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
        ["${"] = Tags.Brace,
        ["$("] = Tags.Paren,
        ["RawString"] = Tags.String,
        ["String"] = Tags.String,
        ["AnsiCString"] = Tags.String,
        ["VariableName"] = Tags.VariableName,
        ["EnvironmentVariable"] = Tags.VariableName,
        ["Functionname"] = Tags.VariableName,
        ["Comment"] = Tags.Comment,
        ["CommandName"] = Tags.Name,
        ["; &"] = Tags.Separator,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var spec = BashParserData.MakeSpec(
            propSources: [BashHighlighting]
        );
        return new LRParser(spec);
    }
}

File: swift/Sources/Rezel/Grammars/BashGrammar.swift
import Foundation

nonisolated(unsafe) let bashHighlighting = styleTags([
	"while do done until for in case esac if then elif else fi": hlTags["controlKeyword"] as Any,
	"IORedirect": hlTags["operator"] as Any,
	"&& || |": hlTags["logicOperator"] as Any,
	"= +=": hlTags["operator"] as Any,
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
	"${": hlTags["brace"] as Any,
	"$(": hlTags["paren"] as Any,
	"RawString": hlString,
	"String": hlString,
	"AnsiCString": hlString,
	"VariableName": hlTags["variableName"] as Any,
	"EnvironmentVariable": hlTags["variableName"] as Any,
	"Functionname": hlTags["variableName"] as Any,
	"Comment": hlComment,
	"CommandName": hlName,
	"; &": hlTags["separator"] as Any,
])

public let bashParser: LRParser = {
	let spec = BashParserData.makeSpec(
		externals: [:],
		propSources: [bashHighlighting]
	)
	return LRParser(spec: spec)
}()
